journal/src/main/scala/com/evolutiongaming/kafka/journal/ConsumerPool.scala

```diff
@@ -0,0 +1,80 @@
+package com.evolutiongaming.kafka.journal
+
+import cats.effect.Resource
+import cats.effect.kernel.{Async, Ref}
+import cats.effect.std.Queue
+import cats.effect.syntax.resource._
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.{Log, LogOf}
+import com.evolutiongaming.kafka.journal.Journals.Consumer
+
+/**
+ * A fixed size pool.
+ *
+ * Semantically blocks on `acquire` if there are no consumers available.
+ *
+ * Make sure to release all consumers before shutting the pool down.
+ */
+private[journal] trait ConsumerPool[F[_]] {
+
+  def acquire: F[Consumer[F]]
+
+  def release(consumer: Consumer[F]): F[Unit]
+}
+
+private[journal] object ConsumerPool {
+
+  case class ConsumerRef[F[_]](consumer: Consumer[F], deallocate: F[Unit])
```

Comment:
I don't see consumer reset (unassign partitions) before returning to pool factored in.

Reply:
I read the javadoc on `org.apache.KafkaConsumer.assign` and it says
```
Manually assign a list of partitions to this consumer. This interface does not allow for incremental assignment and will replace the previous assignment (if there is one).
```

Reply:
> what can happen if we don't unassign partitions before returning to the pool

The records buffered by the consumer will be kept in memory until the next use, the consumer might continue some background work of fetching records behind the scene even if we don't use it anymore. Not sure if there could be any other consequences. The unassign btw might be done on different level - the stream outside of the pool might do it on finalization.

Reply:
I agree that this might be a good idea, but looks like it will lead to lots of code changes. Can we maybe do it in a separate MR?